chain/client/src/view_client.rs

```diff
@@ -758,6 +758,40 @@ impl Handler<GetStateChangesWithCauseInBlock> for ViewClientActor {
     }
 }
 
+/// Returns a hashmap where the key represents the ShardID and the value
+/// is the list of changes in a store with causes for a given block.
+impl Handler<GetStateChangesWithCauseInBlockForTrackedShards> for ViewClientActor {
+    type Result = Result<HashMap<ShardId, StateChangesView>, GetStateChangesError>;
+
+    #[perf]
+    fn handle(
+        &mut self,
+        msg: GetStateChangesWithCauseInBlockForTrackedShards,
+        _: &mut Self::Context,
+    ) -> Self::Result {
+        let state_changes_with_cause_in_block =
+            self.chain.store().get_state_changes_with_cause_in_block(&msg.block_hash)?;
+
+        let mut state_changes_with_cause_split_by_shard_id: HashMap<ShardId, StateChangesView> =
+            HashMap::new();
+        for state_change_with_cause in state_changes_with_cause_in_block {
+            let account_id = &state_change_with_cause.value.get_affected_account_id();
+            let shard_id = match self.runtime_adapter.account_id_to_shard_id(&account_id, &msg.epoch_id) {
```

Comment:
I would expect this to work just fine:

```suggestion
            let account_id = state_change_with_cause.value.get_affected_account_id();
            let shard_id = match self.runtime_adapter.account_id_to_shard_id(account_id, &msg.epoch_id) {
```